src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -1106,6 +1115,116 @@ CreateSchemaDDLCommand(Oid schemaId)
 }
 
 
+/*
+ * GrantOnSchemaDDLCommands creates a list of ddl command for replicating the permissions
+ * of roles on schemas.
+ */
+List *
+GrantOnSchemaDDLCommands(Oid schemaOid)
+{
+	HeapTuple schemaTuple = SearchSysCache1(NAMESPACEOID, ObjectIdGetDatum(schemaOid));
+	bool isNull = true;
+	Datum aclDatum = SysCacheGetAttr(NAMESPACEOID, schemaTuple, Anum_pg_namespace_nspacl,
+									 &isNull);
+	if (isNull)
+	{
+		ReleaseSysCache(schemaTuple);
+		return NIL;
+	}
+	Acl *acl = DatumGetAclPCopy(aclDatum);
+	AclItem *aclDat = ACL_DAT(acl);
+	int aclNum = ACL_NUM(acl);
+	List *commands = NIL;
+	Oid currentUser = GetUserId();
+
+	ReleaseSysCache(schemaTuple);
+
+	for (int i = 0; i < aclNum; i++)
+	{
+		bool withSetRole = (currentUser != aclDat[i].ai_grantor);
+		commands = list_concat(commands,
+							   GenerateGrantOnSchemaQueriesFromAclItem(
+								   schemaOid,
+								   &aclDat[i],
+								   withSetRole));
+	}
+
+	return commands;
+}
+
+
+/*
+ * GenerateGrantOnSchemaQueryFromACL generates a query string for replicating a users permissions
+ * on a schema.
+ */
+List *
+GenerateGrantOnSchemaQueriesFromAclItem(Oid schemaOid, AclItem *aclItem, bool withSetRole)
+{
+	AclMode permissions = ACLITEM_GET_PRIVS(*aclItem) & ACL_ALL_RIGHTS_SCHEMA;
+	AclMode grants = ACLITEM_GET_GOPTIONS(*aclItem) & ACL_ALL_RIGHTS_SCHEMA;
+	Assert(!(grants & ACL_USAGE) || (permissions & ACL_USAGE));
+	Assert(!(grants & ACL_CREATE) || (permissions & ACL_CREATE));
+	Oid granteeOid = aclItem->ai_grantee;
+	List *queries = NIL;
+
+	if (withSetRole)
+	{
+		queries = lappend(queries, GenerateSetRoleQuery(aclItem->ai_grantor));
+	}
+	if (permissions & ACL_USAGE)
+	{
+		char *query = DeparseTreeNode((Node *) GenerateGrantOnSchemaStmtForRights(
+										  granteeOid, schemaOid, "USAGE", grants &
+										  ACL_USAGE));
+		queries = lappend(queries, query);
+	}
+	if (permissions & ACL_CREATE)
+	{
+		char *query = DeparseTreeNode((Node *) GenerateGrantOnSchemaStmtForRights(
+										  granteeOid, schemaOid, "CREATE", grants &
+										  ACL_CREATE));
+		queries = lappend(queries, query);
+	}
+	if (withSetRole)
+	{
+		queries = lappend(queries, "RESET ROLE");
+	}
+	return queries;
+}
+
+
+GrantStmt *
+GenerateGrantOnSchemaStmtForRights(Oid roleOid, Oid schemaOid, char *permission, bool
+								   withGrantOption)
+{
+	GrantStmt *stmt = makeNode(GrantStmt);
+	stmt->is_grant = true;
+	stmt->targtype = ACL_TARGET_OBJECT;
+	stmt->objtype = OBJECT_SCHEMA;
+	stmt->objects = list_make1(makeString(get_namespace_name(schemaOid)));
+	AccessPriv *accessPriv = makeNode(AccessPriv);
+	accessPriv->priv_name = permission;
+	accessPriv->cols = NULL;
+	stmt->privileges = list_make1(accessPriv);
+	RoleSpec *roleSpec = makeNode(RoleSpec);
+	roleSpec->roletype = roleOid ? ROLESPEC_CSTRING : ROLESPEC_PUBLIC;
```

Comment:
`roleOid` is not a boolean, and therefore harder to understand what you are switching on. Could you rewrite these with `OidIsValid(roleOid)`